Remove "Makefile" functionality and update documentation

DanReyLop opened the pull request:
Now that https://github.com/Automattic/wp-calypso/pull/14546 is released and things seem to not being broken, it's time to the "clean up after myself " PR.

* https://github.com/Automattic/wp-calypso/commit/06239f72e6b02f3fd13b1822bed5b3d2d978b1f2 Change documentation to remove references to `make ___` commands, and put `npm run ___` commands instead.
* https://github.com/Automattic/wp-calypso/commit/dc301f52a5108ec4141c58fb7594e2fd36292a54 Remove all the "meat" in the Makefile, leave it as just a bunch of aliases. All the commands you know and love will still work )even `make urn`, that was to good to remove), but they'll just call the equivalent `npm run *` command.
* https://github.com/Automattic/wp-calypso/commit/fa2990f225566e586816eefbc52aa30941ff2614 Remove scripts that are no longer needed (they were only being used by the `Makefile`).

I'll write a P2 shortly with my findings, and some guidelines to keep Calypso working 100% on Windows. Things like "don't assume path separators are `/`", "don't rely on the `#!` shebang working", "don't use shell commands unless you know they work in Windows too" etc.

DanReyLop commented on docs/windows.md:
```diff
@@ -1,51 +0,0 @@
-# Installing Calypso on Windows
```
I'm so happy to be able to delete this file 😂

DanReyLop commented on server/devdocs/bin/generate-components-usage-stats.js:
```diff
@@ -24,15 +24,11 @@ var async = require( 'async' ),
-	// extract list of files to index and remove leading ./'s
 	const outFilePath = 'server/devdocs/components-usage-stats.json';
-	const fileList = globby.sync( process.argv.slice( 2 ) )
-		.map( function( fileWithPath ) {
```
Since the command that invokes this script no longer pipes the output of `find`, removing the initial `./` is no longer needed. This is the only actual code change of this PR.

gwwar commented on Makefile:
```diff
@@ -1,252 +1,62 @@
-	@printf "\033[36m  | (_| (_| | | |_| | |_) \__ \ (_) |    \n"
-	@printf "\033[36m   \___\__,_|_|\__, | .__/|___/\___/     \n"
-	@printf "\033[36m               |___/|_|                  \n"
-	@printf "\033[m\n"
```
;_; save the ascii art

davidhortonweb replied:
♪┏(°.°)┛┗(°.°)┓┗(°.°)┛┏(°.°)┓ ♪

gziolo commented:
Code changes look great 👍  I will try to reproduce the issue @gwwar had.

DanReyLop commented:
> can you see if you can repro this? I'm running into the following when I was testing the docker image.

I can't reproduce it :(

That would only happen if somehow the `docker` image was created without building Calypso (just copying the source code). Have you been experimenting with the Dockerfile lately? If you run `npm run build-docker` before `npm run docker` it shouldn't be possible to have that error (let's hope).

gziolo commented:
I found ` make build` in the docker file. I guess it was updated and then reverted in another PR. We can add it again.

DanReyLop commented:
@gziolo Oh, I made this branch before #14546 was merged, so it doesn't have those changes. But it doesn't really matter, because `make build` will just execute `npm run build` anyway.

gziolo commented:
I removed `server/bundler/assets.json` locally and executed `CALYPSO_ENV=production make build` and it properly created file again, but when I execute ` NODE_ENV=production node build/bundle.js` it fails with:

> Error: Cannot find module 'bundler/assets'
>     at Function.Module._resolveFilename (module.js:469:15)
>     at Function.Module._load (module.js:417:25)
>     at Module.require (module.js:497:17)
>     at require (internal/module.js:20:19)
>     at Object.<anonymous> (/Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/external "bundler/assets":1:1)
>     at __webpack_require__ (/Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/webpack/bootstrap 461607aaef8078ef307f:19:1)
>     at setup (/Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/server/boot/index.js:43:12)
>     at Object.<anonymous> (/Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/index.js:19:8)
>     at __webpack_require__ (/Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/webpack/bootstrap 461607aaef8078ef307f:19:1)
>     at /Users/gziolo/PhpstormProjects/wp-calypso/build/webpack:/webpack/bootstrap 461607aaef8078ef307f:39:1

DanReyLop commented:
@gziolo You scared me there for a moment, I thought I broke everyting. If you run plain `node build/bundle.js` the `NODE_ENV` env var isn't set so it won't find some files. In Docker that's set in this line: https://github.com/Automattic/wp-calypso/blob/master/Dockerfile#L31

gziolo commented:
Sorry about that :)

blowery commented on Makefile:
```diff
@@ -1,252 +1,62 @@
-	@printf "\033[36m  | (_| (_| | | |_| | |_) \__ \ (_) |    \n"
-	@printf "\033[36m   \___\__,_|_|\__, | .__/|___/\___/     \n"
-	@printf "\033[36m               |___/|_|                  \n"
-	@printf "\033[m\n"
```
+1 save the ascii art! 💯

DanReyLop replied:
Ok, ok... by sheer group pressure, I've brought the ASCII art back :)

blowery commented on Makefile:
```diff
@@ -1,252 +1,62 @@
-lint: node_modules/eslint node_modules/eslint-plugin-react node_modules/babel-eslint config-defaults-lint mixedindentlint
-	@$(NPM) run lint
+run:
+	npm start
```
are you excluding the leading `@` intentionally on all these commands?

DanReyLop replied:
No, not intentionally, I didn't even know what it was for. It makes them less verbose, so I've added it to every command.

blowery replied:
👍

gwwar commented:
I cleared out all of my local docker images, rebuilt, and I'm still running into:

![screen shot 2017-06-01 at 10 04 45 am](https://cloud.githubusercontent.com/assets/1270189/26691492/2cea9a30-46b2-11e7-90cb-d02543d8726d.png)

![screen shot 2017-06-01 at 10 08 19 am](https://cloud.githubusercontent.com/assets/1270189/26691514/46048378-46b2-11e7-9d34-15588961564a.png)

gwwar commented:
@DanReyLop okay, that finally cleared itself up after I ran distclean locally. Do you think it'd be worthwhile to add that to the build-docker script?

DanReyLop commented:
> okay, that finally cleared itself up after I ran distclean locally. Do you think it'd be worthwhile to add that to the build-docker script?

I've added `npm run distclean` to the `build-docker` task. Since it's not really gonna be used much in development, it doesn't worry me much, but there should be ways to optimize that Dockerfile :)

I've also rebased with `master`

gwwar commented:
Thanks I've tested this locally + in docker 👍 . 

Could someone test this in Windows? Perhaps @withinboredom, @klimeryk or @nylen ?

withinboredom commented:
I'll take a look at this later today

klimeryk commented on Makefile:
```diff
@@ -1,252 +1,62 @@
-build: install build-server build-css server/devdocs/search-index.js server/devdocs/proptypes-index.json server/devdocs/components-usage-stats.json
-	@if [ $(CALYPSO_ENV) != development ]; then $(BUNDLER); fi
+distclean:
+	@npm run distclean
```
This doesn't seem to be defined:
```
$ npm run distclean

> wp-calypso@0.17.0 distclean /Users/klimeryk/work/workspace/calypso
> npm run -s clean && npm run -s rm -- node_modules

sh: run-p: command not found

npm ERR! Darwin 15.6.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "distclean"
npm ERR! node v6.10.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! wp-calypso@0.17.0 distclean: `npm run -s clean && npm run -s rm -- node_modules`
npm ERR! Exit status 1
npm ERR!
```

DanReyLop replied:
🤦‍♂️ I've never tried to run `clean` with an already empty `node_modules`. You're right, those commands shouldn't need any packages to run. Good catch!

Fixed in https://github.com/Automattic/wp-calypso/pull/14631/commits/76a2b260bcb47e0cfd630f294218c48fb723bf68

klimeryk replied:
Thanks, working nicely now! 👍

nosolosw commented:
FYI, this PR also fixes this issue https://github.com/Automattic/wp-calypso/issues/12511 - which I'm closing because I'm optimistic about this PR :)

nosolosw commented on docs/install.md:
```diff
@@ -7,31 +7,27 @@ You can install Calypso directly on your machine by following the next steps, or
 
 ## Prerequisites
 
 **On Mac OS X or Linux**, to be able to clone the repository and run the application you need:
```
Can we get rid of the `On Mac OS X or Linux` as well?

nosolosw commented on docs/install.md:
```diff
@@ -7,31 +7,27 @@ You can install Calypso directly on your machine by following the next steps, or
 
 **On Mac OS X or Linux**, to be able to clone the repository and run the application you need:
 
 -	Install the [Node.js](http://nodejs.org/) and matching [NPM](https://www.npmjs.com/) [versions](https://nodejs.org/en/download/releases/) listed in the engines section of [package.json](https://github.com/Automattic/wp-calypso/blob/master/package.json). On Mac OS X using [nvm](https://github.com/creationix/nvm) or [n](https://github.com/tj/n) makes it easy to manage `node` versions.
```
Shall we update this to something like? 

- Install the [Node.js](http://nodejs.org/) and matching [NPM](https://www.npmjs.com/) [versions](https://nodejs.org/en/download/releases/) listed in the engines section of [package.json](https://github.com/Automattic/wp-calypso/blob/master/package.json). On Mac OS X and Linux using [nvm](https://github.com/creationix/nvm) or [n](https://github.com/tj/n) makes it easy to manage `node` versions. On Windows you may want to try [nvm-windows](https://github.com/coreybutler/nvm-windows) or [nodist](https://github.com/marcelklehr/nodist).

DanReyLop replied:
Changed this (and your other comment) in https://github.com/Automattic/wp-calypso/pull/14631/commits/57139731abe17bb5f79e78ff188f5a2b914aa053

nosolosw commented on README.md:
```diff
@@ -15,9 +15,8 @@ You can try out the user-side of Calypso on [WordPress.com](https://wordpress.co
-4.	Run `npm install`
-5.	Execute `make run` or `make dashboard` (for a more visually-oriented interface) from the root directory of the repository.
-6.	Open [`calypso.localhost:3000`](http://calypso.localhost:3000/) in your browser.
+4.	Execute `npm start` or `npm run dashboard` (for a more visually-oriented interface) from the root directory of the repository.
```
On a fresh start (without any package installed) `npm run dashboard` doesn't work because `webpack-dashboard` is not installed. Shall we update the `dashboard` command to `"dashboard": "npm install && webpack-dashboard -- npm start",`? This is referenced both in this `README` and the `docs/install.md`

DanReyLop replied:
Fixed: https://github.com/Automattic/wp-calypso/pull/14631/commits/01b497f6be7712c22a887a784b0865114b3ce0ee

nosolosw commented:
Tested a fresh setup (running Calypso first time) on Linux. Other than the comment about `npm run dashboard` this works as expected for that flow.

dmsnell commented on bin/welcome.js:
```diff
@@ -0,0 +1,10 @@
+console.log( chalk.cyan( '   / __/ _\` | | | | | \'_ \\/ __|/ _ \\ ' ) );
+console.log( chalk.cyan( '  | (_| (_| | | |_| | |_) \\__ \\ (_) |  ' ) );
+console.log( chalk.cyan( '   \\___\\__,_|_|\\__, | .__/|___/\\___/ ' ) );
+console.log( chalk.cyan( '               |___/|_|                \n' ) );
```
just saying…this is JS, which should mean two things…

 - why not just use a single multiline template literal?
 - we have access to Unicode, we need more pizazz

DanReyLop replied:
> why not just use a single multiline template literal?

Because I forgot that multiline template literals existed :)

> we have access to Unicode, we need more pizazz

Suit yourself! I don't know how did you folks generate the ASCII art in the first place, but you should be able to tweak it :)

dmsnell replied:
weeeee

`C⃣   a⃣   l⃣   y⃣   p⃣   s⃣   o⃣`

gziolo replied:
We can always open a follow-up to tackle all requests related to Calypso ascii art. It seems like it is the most discussed change here :)

DanReyLop replied:
@dmsnell This is what I see in my machine (Windows 10, Chrome):
![calypso-art](https://user-images.githubusercontent.com/1715800/26905121-5619df8a-4bdd-11e7-949a-d26d3f54ca6b.png)
The ASCII art should be ASCII, specially if we are pushing for compatibility with lesser shells like Windows' `cmd.exe`. In other words: if *I* can't have emojis in my terminal, neither will any of you! 😂

dmsnell replied:
actually @DanReyLop I think this is a major blocker. it's _text_ are and we need emoji. I just don't see how we could move forward without proper gyphage

DanReyLop commented:
So... how do we stand on this? I think this PR is pretty low risk, because the Docker build is already using `npm` anyway. Any objections to merge it? Please don't tell me that the not-awesome-enough ASCII art is gonna be a blocker 😅

gwwar commented:
> So... how do we stand on this?

Should be good to go after someone verifies on Windows. It also wouldn't hurt to smoke test again on Docker from the recent changes.

DanReyLop commented:
> Next step: deprecation message when using make commands? :)

@gziolo I've added a deprecation notice to the `Makefile` itself in https://github.com/Automattic/wp-calypso/pull/14631/commits/803d8fe9dbee9f99fcb6952ce4fca2c06a2d4d60, but the commands will still run cleanly (there's no harm in that). I think the notice is enough. If someone has curiosity, he will see the notice and look in `package.json` for the real commands. And it's enough so no one will keep writing stuff in the Makefile :)

gziolo commented:
Yes, should be enough. At some point everyone is going to love `npm start` command 👍

DanReyLop commented:
> Should be good to go after someone verifies on Windows. It also wouldn't hurt to smoke test again on Docker from the recent changes.

I've just run a smoke test on Docker. I had to fix the `prebuild-docker` command in https://github.com/Automattic/wp-calypso/pull/14631/commits/3541ab41247c3e423020f1e355351dbda1ba98d7 (it was failing in Windows because, after nuking `node_modules`, the `touch` command didn't exist anymore). But otherwise it worked fine.

DanReyLop commented:
> I will say the bash for windows way is the best way. It's almost 500% faster to build for me...

@withinboredom *Maybe* it's because you tried PowerShell first, and so it executed `npm install`? If you tried with `Bash` immediately after (without cleaning or switching branches) then that second build would be much faster because the `npm install` step would be omitted (see `bin/install-if-needed.js`). There *should* be no difference in speed. If anything, `Bash` should be a little slower because of the emulation.

I haven't tried `Bash.exe`, but I don't find the build slow in `cmd.exe`, at least not slower than I remember it in Mac OS X.

Ok, this is it, I'm releasing this :)

withinboredom commented:
@DanReyLop The repos were in two different places (one inside bash filesystem, one in windows filesystem)

DanReyLop commented:
> Bash in windows is actually faster than Windows, because of the emulation. You're talking directly to the kernel (where it's emulated) vs. talking to COM wrappers which talks to COM which talks to the kernel ...

Adding a layer of emulation shouldn't make things faster. I'm not saying you're lying, I'm just saying that if that's true, then it's very sad... and maybe I should try Bash :)

withinboredom commented:
https://blogs.msdn.microsoft.com/commandline/ -- really interesting read! The subsystem is built into the windows kernel. So in some ways, you're bypassing a lot of Windows legacy cruft, which is faster. I've noticed disk bound applications to be "scary fast" in bash.

davidhortonweb commented:
I am also going to update the calypso-bootstrap project to reflect the removal of the makefile as it still references it (https://github.com/Automattic/wp-calypso-bootstrap/issues/11).  Can't wait to start digging in!
